framework/src/headers/ip/v6/icmp/ndp_options.rs

```diff
@@ -0,0 +1,212 @@
+                     byte and bit ordering) is expected to be specified
+                     in specific documents that describe how IPv6
+                     operates over different link layers.  For instance,
+                     [IPv6-ETHER].
+
+   Description
+                     The Source Link-Layer Address option contains the
+                     link-layer address of the sender of the packet.  It
+                     is used in the Neighbor Solicitation, Router
+                     Solicitation, and Router Advertisement packets.
+
+                     The Target Link-Layer Address option contains the
+                     link-layer address of the target.  It is used in
+                     Neighbor Advertisement and Redirect packets.
+
+                     These options MUST be silently ignored for other
+                     Neighbor Discovery messages.
+*/
+#[derive(Debug)]
+#[repr(C, packed)]
+pub struct LinkLayerAddressOption {
+    option_type: u8,
+    option_length: u8,
+    addr: MacAddress,
+}
+
+/// Provides functions to implementors that allow the implementor to scan their NDP message for
+/// NDP options.  The type of options available are specific to the type of NDP Message
+/// Ensures that implementors support NdpMessageContents, which further enforces EndOffset + Sized
+/// which we require to do parsing
+pub trait NdpOptions: NdpMessageContents {
+
+    /// Attempts to lookup the source link layer address from the underlying NDP Message
+    fn get_source_link_layer_address_option(&self, payload_length: u16) -> Option<&MacAddress> {
+        let opt =
+            self.option_scan::<LinkLayerAddressOption>(
+                payload_length,
+                NdpOptionType::SourceLinkLayerAddress
+            );
+        opt.map(|sll| &sll.addr)
+    }
+
+    /// Attempts to lookup the mtu option from the underlying NDP Message
+    fn get_mtu_option(&self, payload_length: u16) -> Option<u32> {
+        let mtu_option = self.option_scan::<MtuOption>(payload_length, NdpOptionType::MTU)
+            .map(|mtu| mtu.get_mtu());
+        mtu_option
+    }
+
+    /// Searches through a packet for an option, returns the first it finds or none
+    /// Relies on the caller to specify T, and they must know that the type can be parsed...
+    /// this is unfortunate, but I do not have a compile time way as the NDPOptionType is something
+    /// like the number 1, and that has to map to a MacAddress, like how do you enforce that?
+    fn option_scan<T>(&self, payload_length: u16, opt_type: NdpOptionType) -> Option<&T> {
+        unsafe {
+            // the start of the options we offset from is the start of the ndp message
+            let opt_start = (self as *const Self) as *const u8;
+            let mut payload_offset = self.offset();
+
+            // make sure we do not scan beyond the end of the packet
+            while (payload_offset as u16) < payload_length {
```

Comment:
this actually doesn't guarantee we don't go past the bounds. we will need to make sure the starting offset for that loop + the expected packet size < total payload length.

Reply:
@drunkirishcoder what do you mean by expected packet size? in the OG code, we would pass in the payload length from the ipv6 header to determine the end of the packet. the ndp options will be the last fields on the ndp messages after the retrans timer. once our pointer is past the retrans timer, we can iterate all the way to the end of the packet

Reply:
So `payload_offset` is the beginning of where the next ndp option starts. This check doesn't cover the scenario where if `(payload_offset + option_len) > payload_len`. In that case you will be overlaying the option struct on top of memory you don't own.

Reply:
added check to make sure the option_length doesnt exceed the payload length